Convert amcrest binary sensors from poll to stream

pnbruckner opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Currently the motion detection binary sensor is polled, which not only causes unnecessary network traffic, it also delays sensor response to motion events. This change converts that sensor to use event streaming from the camera. It also improves the way the online sensor works. Lastly it lays the groundwork for easily adding other types of event driven binary sensors.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->
- Bump amcrest package to 1.6.0.
- For online binary sensor poll camera periodically to test communications in case
  configuration & usage results in no other communication to camera.
- Start a separate thread to call camera's event_stream method since it never returns.
- Convert all received events into signals that cause corresponding sensors to update.
- Use camera's generic event_channels_happened method to update sensors at startup,
  and whenever camera comes back online after being unavailable.

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

pnbruckner committed:
Convert amcrest binary sensors from poll to stream

- Bump amcrest package to 1.6.0.
- For online binary sensor poll camera periodically to test communications in case
  configuration & usage results in no other communication to camera.
- Start a separate thread to call camera's event_stream method since it never returns.
- Convert all received events into signals that cause corresponding sensors to update.
- Use camera's generic event_channels_happened method to update sensors at startup,
  and whenever camera comes back online after being unavailable.

bdraco commented on homeassistant/components/amcrest/__init__.py:
```diff
@@ -184,6 +200,38 @@ def _wrap_test_online(self, now):
 
+def _monitor_events(hass, name, api, event_codes):
+    event_codes = ",".join(event_codes)
+    start_stop = re.compile(r"Code=([^;]+);action=(Start|Stop)", flags=re.S)
```
Can this be parsed without a regex?

balloob replied:
This parsing should be moved into the pypi package.

pnbruckner replied:
There is a _LOT_ that should be moved into the pypi package. In fact, IMHO, for purposes of Home Assistant, it should be pretty much rewritten. But that's not likely to happen anytime soon. Also, Amcrest has changed the API of their cameras and has _not_ documented the changes. They won't even admit that the API has changed, let alone provide any documentation. Newer cameras and newer firmware releases of older cameras are becoming incompatible with the pypi package and Home Assistant. I'm not sure it's worth the effort to make any changes at this point to the pypi package. At least not until Amcrest begins to cooperate again. Is this the only issue with this PR?

bdraco commented on homeassistant/components/amcrest/__init__.py:
```diff
@@ -184,6 +200,38 @@ def _wrap_test_online(self, now):
+    start_stop = re.compile(r"Code=([^;]+);action=(Start|Stop)", flags=re.S)
+    while True:
+        try:
+            api.available_flag.wait()
```
Can wait() throw an exception here?

pnbruckner replied:
No. I suppose it could be outside the `try` block.

bdraco replied:
Have you tested unplugging a camera and blocking all IP traffic to the camera to make sure this doesn’t end up in an infinite loop?

pnbruckner replied:
Yes. When the communication fails, causing the exception, the `api.available_flag` event flag will also be cleared, which will cause it to wait here until the communications start working again (which will result in the flag getting set and the wait will complete.)

pnbruckner replied:
Actually, to be more accurate, when there are a enough communication errors back to back (whether caused here or elsewhere), the flag will get cleared (see `_start_recovery()` above), which will prevent this thread from starting the event stream again until `_wrap_test_online()` detects the camera is back "on line", at which time the flag will be set and this thread will attempt to continue as usual.

bdraco replied:
Should be ok to resolve by moving the `wait()` out of the try

pnbruckner replied:
Will do.

bdraco commented on homeassistant/components/amcrest/helpers.py:
```diff
@@ -1,13 +1,12 @@
-    return signal
+def service_signal(service, *args):
+    """Encode signal."""
+    return slugify("_".join([DOMAIN, service, *args]))
```
I don't think the user is ever going to see this, does this need to be slugifyed ?

pnbruckner replied:
It was an easy way to create a reasonably unique signal name. Is there a problem doing it this way?

bdraco replied:
Does it still work if you take the `slugify` out?


return "_".join([DOMAIN, service, *args])

pnbruckner replied:
Maybe you're missing what this function does. This integration can send a lot of different signals to the various platform modules. This function creates a unqiue signal name based on its parameters. It is only used internal to this integration. I don't see why this is a problem? Have you looked at how it is used within the integration?

balloob replied:
`slugify` is really slow. Frenck and I were able to validate a million entity IDs in 0.6s instead of 12s by no longer using `slugify`: https://github.com/home-assistant/core/pull/32137

pnbruckner replied:
Thanks, and I can understand wanting to optimize. But in this use case this only happens when a camera goes offline or comes back online, or when a motion event is detected. In normal usage I can't see this happening more than once every minute per camera (for motion events. For offline/online it should almost never happen.) So even with six cameras that means not more than once every ten seconds. I measured it with `timeit`, including the regex `findall` call and `slugify`, and it came out to about 20 µsec. So let's say on a really slow machine it's 100 µsec. If 100 µsec / 10 sec is too much for a given installation then I think they have bigger problems to tackle.

As far as why I'm doing this, I was attempting to make the signal names be somewhat uniform. One of the items that goes into it is the camera's name (not its object ID.) But I guess there aren't any restrictions on what a signal name can be, just that they're unique. I already guarantee uniqueness of camera names, so I suppose I don't really need `slugify`. I'll remove it.

bdraco reviewed (changes requested):
Everything else looks ok

pnbruckner committed:
Changes per review

bdraco commented on homeassistant/components/amcrest/helpers.py:
```diff
@@ -1,13 +1,12 @@
-    return signal
+def service_signal(service, *args):
+    """Encode signal."""
+    return slugify("_".join([DOMAIN, service, *args]))
```
```suggestion
    return "_".join([DOMAIN, service, *args])
```

`slugify` is doing this under the hood:
```
re.sub(r'[-\s]+', '-',
            unicode(
                re.sub(r'[^\w\s-]', '',
                    unicodedata.normalize('NFKD', string)
                    .encode('ascii', 'ignore'))
                .strip()
                .lower()))
```

```
>>> from homeassistant.components.amcrest.helpers import service_signal
>>> from homeassistant.components.amcrest.const import SERVICE_UPDATE, DOMAIN
>>> service_signal(SERVICE_UPDATE,"name","code")
'amcrest_update_name_code'
>>> "_".join([DOMAIN, SERVICE_UPDATE, "name", "code"])
'amcrest_update_name_code'
```

I'm missing why its needed here

pnbruckner replied:
Yes, thank you, but I know what `slugify` does. I've looked at its code before.

This will give you an idea why I was using it:
```python
>>> from homeassistant.components.amcrest.const import DOMAIN, SERVICE_EVENT, SERVICE_UPDATE
>>> from homeassistant.components.amcrest.helpers import service_signal
>>> service_signal(SERVICE_UPDATE, "Living Room IPC")
'amcrest_update_living_room_ipc'
>>> "_".join([DOMAIN, SERVICE_UPDATE, "Living Room IPC"])
'amcrest_update_Living Room IPC'
>>> service_signal(SERVICE_EVENT, "Living Room IPC", "VideoMotion")
'amcrest_event_living_room_ipc_videomotion'
>>> "_".join([DOMAIN, SERVICE_UPDATE, "Living Room IPC", "VideoMotion"])
'amcrest_update_Living Room IPC_VideoMotion'
```
But as I said above, I'll remove it.

pnbruckner committed:
Changes per review 2

pnbruckner commented:
@bdraco is this acceptable now? Can you give your approval, or is there something else you'd like changed? Thanks!

bdraco commented:
> @bdraco is this acceptable now? Can you give your approval, or is there something else you'd like changed? Thanks!

I think the only outstanding item is @balloob 's comment about moving the regex into the pypi package and then I'm good to signoff

pnbruckner commented:
In principal I agree the parsing should be done in the pypi package.

However, it seems a bit arbitrary to want to require it for this but not for all the other camera/sensor commands that have to do extra processing in the HA integration that should have been done in the pypi package. The problem is that the package is used by more than just HA, and changing its API is problematic. Also, the package is not very well written in the first place and, as I said, should be completely rewritten/replaced for HA usage. But nobody is signing up to do that. (I may be the current code owner of this integration, but I inherited it, I didn't write it. And have only contributed to the pypi package. I'm not its owner.)

It works as it is, and it seems to me this is just delaying getting an often requested improvement to the users of this integration. I would request it be accepted as is. Maybe someday I or someone else will have the time to rewrite/replace the package.

bdraco commented:
> In principal I agree the parsing should be done in the pypi package.
> 
>  (I may be the current code owner of this integration, but I inherited it, I didn't write it. And have only contributed to the pypi package. I'm not its owner.)
> 
Ok lets go for a different approach then.

How about moving the monitor into a new file say monitor.py so its more isolated from the existing code?

Also please store the refcount in hass.data[DOMAIN]..... instead of a global.

pnbruckner commented:
> How about moving the monitor into a new file say monitor.py so its more isolated from the existing code?

That seems like make work. I don't see the point. Why should it be isolated from the module that uses it? That would not solve the original issue that the pypi package doesn't isolate the user (in this case the HA integration) from low level details. As I said, this is not the only example of code in the HA integration that belongs in the pypi package. It it's going to be reorganized, then to me that's a separate task that at least belongs in a different PR.

> Also please store the refcount in hass.data[DOMAIN]..... instead of a global.

That's not right. The data is only used within the one module. `hass.data` is for sharing data across modules. A module-specific piece of data should not be shared publicly.

I hate to sound argumentative, but I don't think these are appropriate requests.

bdraco commented:
@pnbruckner I'm going to step out on this one.  I think this is best handled by another reviewer.

frenck commented on homeassistant/components/amcrest/__init__.py:
```diff
@@ -184,6 +200,38 @@ def _wrap_test_online(self, now):
+
+
+def _start_event_monitor(hass, name, api, event_codes):
+    global _event_threads  # pylint: disable=invalid-name
```
```suggestion
    global _event_threads  # pylint: disable=invalid-name
```

@bdraco is right here, don't use global.

pnbruckner replied:
Can someone explain why?

frenck replied:
It is more of a generally programming concept. I'm pretty darn sure if you search for "Why are globals bad/evil" you will find tons of good reasons.

Don't get me wrong, we have some globals in the core of Home Assistant, they have (rare) use cases; Other than that, we aren't using it anywhere.

pnbruckner replied:
Thanks, but I've been a programmer for a long time. I know the trade-offs of globals. Just because they can be bad in some or even most uses cases doesn't make them bad in all. But if globals shouldn't be used, then isn't `hass.data` effectively a global, too? And a very public one at that? Much more public than a module level global with a leading underscore. I fail to see why that would be better. What am I missing?

As a compromise, would it be better if I converted those functions into a Class (possibly a subclass of `threading.Thread`), turning the global into a class variable?

frenck replied:
> Thanks, but I've been a programmer for a long time. I know the trade-offs of globals.

Sorry m8, it was not a personal attack in any way. I just meant to answer a question that consisted of 4 words. I cannot read or extract any level of knowledge or other possible emotions from that. So forgive me for providing a general answer, I mean no harm.

`hass` isn't global, although I understand the common problem you are raising in this use case. The difference here would be that `hass.data[DOMAIN]` is specially meant for storing run-time application data (in a sense globally, but in a single place, namespaced, and passed in and documented). IMO is better than the growth of globally defined variables everywhere throughout the codebase.

However, that can become a long opinionated discussion 😉 

> As a compromise, would it be better if I converted those functions into a Class (possibly a subclass of threading.Thread), turning the global into a class variable?

Yeah, that might work, handling the monitoring of events in a separate class.

balloob replied:
We should use hass.data over global. That way we can guarantee that things are global on a per-hass instance basis.

However, in this case, why not use the `name` arg as part of the name ? That would be the most recognizable I think.

balloob replied:
Some background about the `hass.data` instead of `global` decision: our testing infrastructure would end up relying on all these accidental left around globals, would break if the test order would change and leak memory. It was a lot of pain but we got rid of most globals. 

You will still find some globals but those should be cleaned up one day. It's the same with asking people to put things in PyPI packages. We first need to stop it from happening, then we can start reducing it to eventually eliminate it. 

We have blocked pull requests because integrations first needed to have existing code be put into a PyPI package before we were able to accept new pull requests for it. 

The goal of the PyPI rule is for reusability and maintainability. PyPI packages can be reused by other projects which will result in improved quality and less bugs, flourishing the open source ecosystem. The 2nd reason is that reviewers are really overloaded and we should not have to go through PRs that contain protocol specific information but instead focus on the integration of it in Home Assistant.

In this case, I think that the `event_stream` method should emit event objects/dictionaries that can be consumed by Home Assistant. This will improve the re-usability of the library. 

I see that the [`amcrest` package](https://github.com/tchellomello/python-amcrest) was started by Marcelo. I know him personally and also recently contributed to his Ring package. Pretty sure that there will be a quick turnaround time.

pnbruckner replied:
@balloob, thanks for the explanation. That makes a lot of sense.

Yes, I could use the `name` arg in constructing the thread's name, which would obviate the need for a counter. Since these names are not constrained in any way (other than being unique from other amcrest camera names), can I use them as-is (e.g., "`Amcrest Kitchen IPC`"), or should I massage them in some way? I.e., would you want the thread names to meet some sort of naming convention? FWIW, this processing would only need to happen once during startup for each camera.

Regarding the amcrest package, I was given write permission quite some time ago and have done most of the recent work, especially for use by Home Assistant. I could certainly push more of the processing down into the package where it belongs. But as I've said, if I do it for this, then I should really be doing it for all the other commands/methods used by HA. And I would have to do it in such a way as to maintain backwards compatibility for all the other use cases of that package. E.g., I'd probably have to add a layer on top of most of the existing commands, but that just seems like a band-aid. And there are several other major changes that should probably be made (such as not embedding communications in property methods, etc.) I've considered forking the project and doing a major overhaul to bring it more in line with HA expectations, dropping Python2 support, cli, etc., but that is a significant project. Although I agree it's not optimal, is it ok to leave this as-is for now, especially since it is consistent with the way the rest of the package is used, and then maybe tackle the overhaul later?

balloob replied:
We have no rules for thread names, no massaging necessary.

I think that it's not necessary to fork the project and revamp all CLI commands just to integrate this. I would just do a breaking change bump to 2.0 and just include this one thing.

pnbruckner committed:
Changes per review 3

- Move event stream decoding to amcrest package.
- Change name of event processing threads so global
  counter is no longer required.
- Bump amcrest package to 1.7.0.
